Polars: add `SeriesDomain`

Shoeboxam opened the pull request:
Closes #762

Shoeboxam commented:
Current dependencies on/for this PR:
* main
  * **PR #763** <a href="https://app.graphite.dev/github/pr/opendp/opendp/763" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/opendp/opendp/763?utm_source=stack-comment).

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+    type Polars = StringType;
+}
+#[cfg(test)]
+mod test_series {
```
It would be really good to get Rust test coverage working. I can see names up above that aren't referenced here... but maybe they are used indirectly? It's hard for me to know.

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+
+    #[test]
+    fn test_series_new() -> Fallible<()> {
+        let series_domain = SeriesDomain::new("A", AtomDomain::<bool>::default());
+
+        let series = Series::new("A", vec![true; 50]);
+        assert!(series_domain.member(&series)?);
+        assert!(series_domain == series_domain);
```
Move `assert!(series_domain == series_domain);` up to immediately after definition.

```suggestion
        let series_domain = SeriesDomain::new("A", AtomDomain::<bool>::default());
        assert!(series_domain == series_domain);
        
        let series = Series::new("A", vec![true; 50]);
        assert!(series_domain.member(&series)?);
```

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+    fn test_series_bounded() -> Fallible<()> {
+        let series_domain = SeriesDomain::new("A", AtomDomain::new_closed((1, 3))?);
+
+        let series = Series::new("A", vec![1; 50]);
+        assert!(series_domain.member(&series)?);
+
+        let series = Series::new("A", vec![4; 50]);
+        assert!(!series_domain.member(&series)?);
```
```suggestion
        let inside_bounds = Series::new("A", vec![1; 50]);
        assert!(series_domain.member(&inside_bounds)?);

        let outside_bounds = Series::new("A", vec![4; 50]);
        assert!(!series_domain.member(&outside_bounds)?);
```

The `!` is hard to spot: names can make intent more clear.

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+            assert!(series_domain.member(&series)?);
+        }
+
+        // nullable type without options
```
Maybe split each case into a separate test? Or use different names for each case, rather than just `series_domain`.

mccalluc commented on rust/src/error.rs:
```diff
@@ -141,6 +154,16 @@ impl From<ConversionError> for Error {
+            backtrace: std::backtrace::Backtrace::capture(),
+        }
+    }
+}
```
Test the changes in error handling?

mccalluc reviewed (approved):
- Rust test coverage would be nice
- Preference for either smaller tests, or using variable names that reflect the case being tested

Approved: Nothing that needs to be handled right now.

Shoeboxam commented on rust/src/domains/mod.rs:
```diff
@@ -11,6 +11,9 @@
 pub(crate) mod ffi;
 
+mod polars;
+pub use crate::domains::polars::*;
```
Yep!

Shoeboxam commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+/// let series = SeriesDomain::new("A", AtomDomain::<i32>::default());
+/// // Create a SeriesDomain with column `B` and `f64` AtomDomain with bounds `[1.0, 2.0]`.
+/// let series_with_bounds = SeriesDomain::new("B", AtomDomain::<f64>::new_closed((1.0, 2.0))?);
+/// # opendp::error::Fallible::Ok(())
```
Yes, Rust implicitly wraps it in a function. The last line is behind a hash to hide it in the rendered documentation.

Shoeboxam commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,307 @@
+            DataType::Float32 => atom_member!(f32, Float32Type),
+            DataType::Float64 => atom_member!(f64, Float64Type),
+            DataType::Boolean => atom_member!(bool, BooleanType),
+            DataType::String => atom_member!(str, StringType),
```
Nope, the arrow spec doesn't include types that are specific to machine architecture.

mccalluc reviewed (dismissed):
Having trouble seeing the exact changes since last review, but seems ok.

Shoeboxam commented:
Changes in this PR since last review:
1. pulled all dependencies down into the Cargo.toml
1. made polars an optional feature that is not enabled by default
2. added the polars feature to CI
3. addressed or responded to review comments

Sorry the diffs aren't in a new commit.

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,315 @@
+        let series_domain = SeriesDomain::new("A", AtomDomain::<f64>::new_nullable());
+
+        // None is not ok, but NaN is ok
+        let series = Series::new("A", vec![Some(1.), Some(f64::NAN), None]);
+        assert!(!series_domain.member(&series)?);
+        // series made with Option::Some are ok
+        let series = Series::new("A", vec![Some(1.), Some(f64::NAN)]);
+        assert!(series_domain.member(&series)?);
+        // series made without options are ok
+        let series = Series::new("A", vec![1., f64::NAN]);
+        assert!(series_domain.member(&series)?);
```
```suggestion
        let series_with_none = Series::new("A", vec![Some(1.), Some(f64::NAN), None]);
        assert!(!series_domain.member(&series_with_none)?);
        
        // series made with Option::Some are ok
        let series_with_some = Series::new("A", vec![Some(1.), Some(f64::NAN)]);
        assert!(series_domain.member(&series_with_some)?);
        
        // series made without options are ok
        let series_wo_some = Series::new("A", vec![1., f64::NAN]);
        assert!(series_domain.member(&series_wo_some)?);
```

Where we really have separate tests that happen to be run serially, I think it improves readability to use distinct names, and add a line break between.

mccalluc commented on rust/src/domains/polars/series/mod.rs:
```diff
@@ -0,0 +1,315 @@
+        let series_domain =
+            SeriesDomain::new("A", OptionDomain::new(AtomDomain::<f64>::new_nullable()));
+
+        // None and NaN are both ok
+        let series = Series::new("A", vec![Some(1.), Some(f64::NAN), None]);
+        assert!(series_domain.member(&series)?);
+        // doesn't have to have NaN
+        let series = Series::new("A", vec![1., 2.]);
+        assert!(series_domain.member(&series)?);
```
```suggestion
        // None and NaN are both ok
        let series_with_none = Series::new("A", vec![Some(1.), Some(f64::NAN), None]);
        assert!(series_domain.member(&series_with_none)?);
        
        // doesn't have to have NaN
        let series_wo_none = Series::new("A", vec![1., 2.]);
        assert!(series_domain.member(&series_wo_none)?);
```
ditto

mccalluc reviewed (approved):
A couple suggestions for test readability, but doesn't need to hold up the show.

Shoeboxam commented:
Added suggestions. One more change: when all deps are present, there's not enough disk space in the runner. I added a step to smoke-test to clear unused software that comes with the image. Will need something more for building on other platforms in release.

mccalluc commented on .github/workflows/smoke-test.yml:
```diff
@@ -16,6 +16,18 @@ jobs:
     steps:
+
+      - name: Free disk space
+        uses: jlumbroso/free-disk-space@main
```
```suggestion
        uses: jlumbroso/free-disk-space@v1.3.1
```
Pin version?

mccalluc commented on .github/workflows/smoke-test.yml:
```diff
@@ -37,13 +49,13 @@ jobs:
 
       - name: Build
-        run: cargo build --verbose --features untrusted,bindings
+        run: cargo build --verbose --features untrusted,bindings,polars
```
Background question: What is gained by separating the polars feature? Is it for Rust users who might want only a subset of functionality? Is it to keep developers from accidentally adding dependencies to libraries where they aren't really needed?

We have some cleanup to do on the existing features (#1343, #1338), so I'd like to understand what we gain.

Shoeboxam replied:
1. R will not support Polars out-of-the-box, so R users don't need to compile it (it's a big dep)
2. Users can disable Polars and get a WebAssembly build once OpenSSL is dropped
3. We were thinking of initially making this part of a patch release, where it would be better to leave it off by default. However, there are some other breaking changes in main, so this is not applicable.

mccalluc commented on python/src/opendp/_extrinsics/_make_np_eigenvector/__init__.py:
```diff
@@ -75,7 +75,7 @@ def function(C):
             #       sample with sufficient precision where all components round to same float
             
-            z = np_csprng.multivariate_normal(mean=np.zeros(d), cov=Omega_inv)
+            z = np_csprng.multivariate_normal(mean=np.zeros(d), cov=Omega_inv)  # type: ignore[union-attr] 
```
Is this necessary? I don't see changes in this PR that are obviously related.

Shoeboxam replied:
Strange, my MyPy fails if I don't have it.

mccalluc commented on rust/Cargo.toml:
```diff
@@ -42,6 +42,28 @@ openssl = { version = "0.10.29", features = ["vendored"], optional = true }
 opendp_tooling = { path = "opendp_tooling", optional = true, version = "0.9.2-dev" }
 readonly = "0.2"
 
+polars = { version = "=0.38.3", features = [
```
Is there a reason for specifying the version with or with-out `=`? Add a comment to explain, or pick one or the other and be consistent?

mccalluc reviewed (approved):
Not sure if there are particular things to re-review, but looks ok after spot check.
